cirq-core/cirq/transformers/diagonal_optimization.py

```diff
@@ -0,0 +1,134 @@
+    later in the circuit. This handles cases where diagonal gates can commute past
+    other operations. For example:
+
+        Z(q0) - CZ(q0, q1) - measure(q1)
+
+    After `eject_z`, the Z gate on the control qubit commutes through the CZ:
+
+        CZ(q0, q1) - Z(q1) - measure(q1)
+
+    Then both the CZ and Z(q1) can be removed since they're before the measurement:
+
+        measure(q1)
+
+    Args:
+        circuit: Input circuit to transform.
+        context: `cirq.TransformerContext` storing common configurable options for transformers.
+
+    Returns:
+        Copy of the transformed input circuit with diagonal gates before measurements removed.
+
+    Examples:
+        >>> import cirq
+        >>> q0, q1 = cirq.LineQubit.range(2)
+        >>>
+        >>> # Simple case: Z before measurement
+        >>> circuit = cirq.Circuit(cirq.H(q0), cirq.Z(q0), cirq.measure(q0))
+        >>> optimized = cirq.drop_diagonal_before_measurement(circuit)
+        >>> print(optimized)
+        0: ───H───M───
+
+        >>> # Complex case: Z-CZ commutation
+        >>> circuit = cirq.Circuit(
+        ...     cirq.Z(q0),
+        ...     cirq.CZ(q0, q1),
+        ...     cirq.measure(q1)
+        ... )
+        >>> optimized = cirq.drop_diagonal_before_measurement(circuit)
+        >>> print(optimized)
+        1: ───M───
+    """
+    if context is None:
+        context = transformer_api.TransformerContext()
+
+    # Phase 1: Push Z gates later in the circuit to maximize removal opportunities.
+    circuit = eject_z(circuit, context=context)
+
+    # Phase 2: Remove diagonal gates that appear before measurements.
+    # We iterate in reverse to identify which qubits will be measured.
+    # Track qubits that will be measured (set grows as we go backwards)
+    measured_qubits: set[ops.Qid] = set()
+
+    # Build new moments in reverse
+    new_moments = []
+    for moment in reversed(circuit):
+        new_ops = []
+
+        for op in moment:
+            # If this is a measurement, mark these qubits as measured
+            if protocols.is_measurement(op):
+                measured_qubits.update(op.qubits)
+                new_ops.append(op)
+            # If this is a diagonal gate and ANY of its qubits will be measured, remove it
+            # (diagonal gates only affect phase, which doesn't impact computational basis
+            # measurements)
+            elif _is_diagonal(op):
+                # CRITICAL: we can only remove if all qubits involved are measured.
+                # if even one qubit is NOT measured, the gate must stay to preserve
+                # the state of that unmeasured qubit (due to phase kickback/entanglement).
+                if all(q in measured_qubits for q in op.qubits):
+                    continue  # Drop the operation
+
+                new_ops.append(op)
+                # Note: We do NOT remove qubits from measured_qubits here.
+                # Diagonal gates commute with other diagonal gates.
```

Comment:
Can you add a test for this? A CZ gate followed by a diagonal 4x4 unitary followed by measurement on both qubits.